### PR TITLE
I496 skew join

### PR DESCRIPTION
**Pull request by laneb**
Addressing #496.

### CODE REVIEW

**Comment by laneb** [comment]
Build failed on an unrelated test due to a lost executor. Hope that isn't a recurring issue. There is a button to restart the build on the detail page.

File: src/main/python/smv/helpers.py [hunk 1]
```diff
@@ -341,6 +341,14 @@ def smvJoinMultipleByKey(self, keys, joinType = 'inner'):
         jdf = self._jPythonHelper.smvJoinMultipleByKey(self._jdf, smv_copy_array(self._sc, *keys), joinType)
         return SmvMultiJoin(self._sql_ctx, jdf)
 
+    def topNValsByFreq(self, n, col):
```
Comment (AliTajeldin) [inline src/main/python/smv/helpers.py]: Missing pydoc.

File: src/main/python/smv/helpers.py [hunk 2]
```diff
@@ -341,6 +341,14 @@ def smvJoinMultipleByKey(self, keys, joinType = 'inner'):
...
+        topNdf = DataFrame(self._jDfHelper._topNValsByFreq(n, col._jc), self._sql_ctx)
+        return map(lambda r: r.asDict().values()[0], topNdf.collect())
+
+    def smvSkewJoinByKey(self, other, joinType, skewVals, key):
```
Comment (AliTajeldin) [inline src/main/python/smv/helpers.py]: missing pydoc

File: src/main/scala/org/apache/spark/sql/contrib/smv/package.scala [hunk 3]
```diff
@@ -62,4 +65,13 @@ package object smv {
     val converter = CatalystTypeConverters.createToCatalystConverter(dt)
     converter(sv)
   }
+
+  def hasBroadcastHint(df: DataFrame): Boolean =
```
Comment (AliTajeldin) [inline src/main/scala/org/apache/spark/sql/contrib/smv/package.scala]: needs comment to describe what this method is doing and more importantly, the why.

File: src/main/scala/org/apache/spark/sql/contrib/smv/package.scala [hunk 4]
```diff
@@ -62,4 +65,13 @@ package object smv {
...
+    df.logicalPlan match {
+      case _: BroadcastHint => true
+      case _                => false
+    }
+
+  def propagateBroadcastHint(dfParent: DataFrame, dfChild: DataFrame): DataFrame =
```
Comment (AliTajeldin) [inline src/main/scala/org/apache/spark/sql/contrib/smv/package.scala]: Missing description.  We will forget why we added this in a year.

File: src/main/scala/org/tresamigos/smv/SmvDFHelper.scala [hunk 5]
```diff
@@ -389,7 +394,7 @@ class SmvDFHelper(df: DataFrame) {
    *
    * and the following call:
    * {{{
-   *   df.dedupByKey("id")
+   *   df.debupByKey("id")
```
Comment (AliTajeldin) [inline src/main/scala/org/tresamigos/smv/SmvDFHelper.scala]: Here is the `debupByKey` again.  why?  Should be `dedupByKey`

File: src/main/scala/org/tresamigos/smv/SmvDFHelper.scala [hunk 6]
```diff
@@ -811,6 +816,34 @@ class SmvDFHelper(df: DataFrame) {
...
+  /*
+   * Create single-columned dataframe whose values are the top N for the specified
+   * column. For the Python port, we cannot return the sequence of values across
+   * Py4J because the values have type Any and Py4J does not how to cast them as
+   * Python values. Therefore, we collect the DataFrame into a list in Python
+   */
+  private[smv] def _topNValsByFreq(n: Integer, c: Column) =
+    df.groupBy(c).agg(count(c) as "freq").smvTopNRecs(n, col("freq").desc).select(c)
+
+  def topNValsByFreq(n: Integer, c: Column) =
```
Comment (AliTajeldin) [inline src/main/scala/org/tresamigos/smv/SmvDFHelper.scala]: Missing scaladoc.

File: src/main/scala/org/tresamigos/smv/SmvDFHelper.scala [hunk 7]
```diff
@@ -811,6 +816,34 @@ class SmvDFHelper(df: DataFrame) {
...
+    _topNValsByFreq(n, c).collect() map (_.get(0))
+
+  def smvSkewJoinByKey(df2: DataFrame,
```
Comment (AliTajeldin) [inline src/main/scala/org/tresamigos/smv/SmvDFHelper.scala]: Missing scaladoc.  This is a public interface.

**Comment by laneb** [comment]
There should be an automatic Python API doc including these methods very soon. Still would probably be helpful to add this to the user doc.

**Pull request merged**